riot_pal/serial_driver.py

```diff
@@ -65,18 +71,19 @@ def readline(self):
         """Read and decode to utf-8 data.
 
         Returns:
-            str: string of data if success, ERR string if failed.
+            str: string of data if success, empty string if failed.
         """
         try:
             res_bytes = self._dev.readline()
             response = res_bytes.decode("utf-8", errors="ignore")
         except (ValueError, TypeError, SerialException) as exc:
-            response = 'ERR'
+            response = ''
             logging.debug(exc)
-        if response == '':
-            logging.debug("Reconnecting due to timeout")
-            self.close()
-            self._connect(*self.args, **self.kwargs)
+        else:
```

Comment:
this changes the behaviour a bit, so reconnect will only happen if there was **no** error (i.e., not going into except above). You want that?

Reply:
Not really, since a TypeError would cause the response to be set to "ERR" which was just not an empty string it would not reconnect.

Since I didn't want this arbitrary "ERR" I added the except else stuff.  Functionality should be the same it just won't return "ERR" anymore.  This can be tested with robot tests since we try to timeout.

Reply:
not sure, before there was an `if` which is not part of `try except` and was hence executed every time, i.e. on success and error. However the `else` now is part of `try except` and that branch will only run if there was no error see https://files.realpython.com/media/try_except_else.703aaeeb63d3.png

But from you description you want to go into the `if` branch on an empty `response`, right?

Reply:
Yup, only if nothing was received.  If it was a parsing or decoding error I don't think it would require a reconnect.

line 74 response = 'ERR' means that it would never reconnect on exception.

I think this change makes that a bit more clear that my intention is to only reconnect if a successful timeout has occured...

Reply:
okay, I think I understand now 😄